internal/gps/source_cache_test.go

```diff
@@ -0,0 +1,465 @@
+					}
+				})
+			}
+		})
+	})
+}
+
+// compareManifests compares two manifests and reports differences as test errors.
+func compareManifests(t *testing.T, exp, got Manifest) {
+	{
+		exp, got := exp.DependencyConstraints(), got.DependencyConstraints()
+		if !projectConstraintsEqual(exp, got) {
+			t.Errorf("expected constraints equal to %v but got %v", exp, got)
+		}
+	}
+	{
+		exp, got := exp.TestDependencyConstraints(), got.TestDependencyConstraints()
+		if !projectConstraintsEqual(exp, got) {
+			t.Errorf("expected test constraints equal to %v but got %v", exp, got)
+		}
+	}
+
+	expRM, expOK := exp.(RootManifest)
+	gotRM, gotOK := got.(RootManifest)
+	if expOK && !gotOK {
+		t.Errorf("expected RootManifest but got %v", got)
+		return
+	}
+	if gotOK && !expOK {
+		t.Errorf("didn't expected RootManifest: %v", got)
+		return
+	}
+
+	{
+		exp, got := expRM.IgnoredPackages(), gotRM.IgnoredPackages()
+		if !mapStringBoolEqual(exp, got) {
+			t.Errorf("expected ignored packages equal to %v but got %v", exp, got)
+		}
+	}
+
+	{
+		exp, got := expRM.Overrides(), gotRM.Overrides()
+		if !projectConstraintsEqual(exp, got) {
+			t.Errorf("expected overrides equal to %v but got %v", exp, got)
+		}
+	}
+
+	{
+		exp, got := expRM.RequiredPackages(), gotRM.RequiredPackages()
+		if !mapStringBoolEqual(exp, got) {
+			t.Errorf("expected required packages equal to %v but got %v", exp, got)
+		}
+	}
+}
+
+// compareLocks compares two manifests and reports differences as test errors.
+func compareLocks(t *testing.T, exp, got Lock) {
+	if !bytes.Equal(exp.InputHash(), got.InputHash()) {
+		t.Errorf("expected lock input has %s but got %s", exp.InputHash(), got.InputHash())
+	}
+	{
+		exp, got := exp.Projects(), got.Projects()
+		SortLockedProjects(exp)
+		SortLockedProjects(got)
+		if len(exp) != len(got) {
+			t.Errorf("expected %d locked projects but got %d: expected %v\ngot %v", len(exp), len(got), exp, got)
+		}
+		for i := range exp {
+			if !exp[i].Eq(got[i]) {
+				t.Errorf("index %d: expected locked project equal to %v but got %v", i, exp[i], got[i])
+			}
+		}
+	}
+}
+
+// comparePackageTree compares two pkgtree.PackageTree and reports differences as test errors.
+func comparePackageTree(t *testing.T, exp, got pkgtree.PackageTree) {
+	if got.ImportRoot != exp.ImportRoot {
+		t.Errorf("expected package tree root %q but got %q", exp.ImportRoot, got.ImportRoot)
+	}
+	{
+		exp, got := exp.Packages, got.Packages
+		if len(exp) != len(got) {
+			t.Errorf("expected packages %v but got %v", exp, got)
+		} else {
+			for k, v := range exp {
+				if v2, ok := got[k]; !ok {
+					t.Errorf("key %q: expected %v but got none", k, v)
+				} else if !packageOrErrEqual(v, v2) {
+					t.Errorf("key %q: expected %v but got %v", k, v, v2)
+				}
+			}
+		}
+	}
+}
+
+// pairedVersionsEqual returns true if the PairedVersions have equal VersionTypes and Strings.
+func pairedVersionsEqual(a, b PairedVersion) bool {
+	return a.Type() == b.Type() && a.String() == b.String()
```

Comment:
I have some ideas to make this one more robust (at least we control all the implementing types, so worst case it is just cumbersome). I'll try to come up with some specific cases that break this approach first though.